### ISSUE TITLE
Non-watch Webpack instances hang when dashboard is running

### ISSUE THREAD

**Issue by aequasi** [FormidableLabs/webpack-dashboard#96]
My workflow has a watch, and then a non-watch webpack process. The non-watch one runs any time I push or pull from git. When i have the watch process running, with the dashboard, the non-watch process hangs until i kill it or the dashboard.
# 
###### Steps to reproduce the problem

Run any webpack profile, without the `--watch` flag, and with the plugin included, and it will hang after its finished, with or without webpack-dashboard actually running.
# 
###### Please provide a gist of relevant files

`webpack -d --progress --colors --config webpack.config.js`
Webpack Config File https://gist.github.com/aequasi/7cc2378fc965c013ade0b4a1a4f6a8a5
# 
###### More Details
- Any OS
- ITerm2/Terminal/PHPStorm
- 0.1.1
- xterm-256color

**Comment by jamesbvaughan** [comment]
@aequasi What's the advantage of having the non-watch webpack for git stuff?

**Comment by aequasi** [comment]
when i push and pull, it compiles the assets (before and after, respectively) automatically

**Comment by zhaozhiming** [comment]
I have a webpack-dashboard hang problem too.

When I run dev env with script is work fine.
`CLIENT_PORT=4000 NODE_ENV=development webpack-dashboard -- webpack-dev-server --inline --config webpack/client.js`

But when I run pro env script with follow is hang.
`NODE_ENV=production webpack --config webpack/client.js --progress`

My github project: https://github.com/zhaozhiming/sifancai

**Comment by kenwheeler** [comment]
This is because you both are including the dashboard plugin in production. Putting a flag around it so it only gets added in dev, or splitting your config files into dev and prod should fix it.

**Comment by mariusbalcytis** [comment]
Yes, but is there a reason for plugin to "hang" the process?

This makes it somewhat more difficult to keep configurations intact - you cannot call `webpack` with `webpack-dashboard` if plugin is not enabled and if you do enable it, it will hang the process even if it's called without `webpack-dashboard`.

Could it maybe look at the child process? As `webpack` exits after job is done and `webpack-dev-server` does not - just like `webpack-dashboard` should when running one or another.

**Comment by aequasi** [comment]
Yeah, i dont see why this should hang a process thats running without --watch

**Comment by kenwheeler** [comment]
You guys don't separate your webpack configs into dev and production?

**Comment by kenwheeler** [comment]
It hangs because it opens a socket presumably

**Comment by aequasi** [comment]
I don't need to separate my configs. Yes, i would assume it's hanging because it's waiting on communication from the sockets

**Comment by kenwheeler** [comment]
So why not wrap it with a prod conditional like you do with uglify

**Comment by aequasi** [comment]
Thats what i resorted to doing, but it shouldnt hang if its not a watch, regardless

**Comment by kenwheeler** [comment]
Not sure of how I'd go about detecting that it's not a watch

**Comment by fancyfootowork** [comment]
I ran into the same error. 
I ended up adding yargs. Seems to do the trick:

`const argv = require('yargs').argv;`
then after setting up the main config:
```
if (argv.watch) {
  const DashboardPlugin = require('webpack-dashboard/plugin');
  config = merge(config, { plugins: [new DashboardPlugin()] });
}
```

**Comment by sethbro** [comment]
:+1 Seems wrong to hang when invoking `webpack` without `webpack-dashboard`.

@fancyfootowork's yargs solution worked for me.

**Comment by MarkPieszak** [comment]
@sethbro I tried the yargs solution but got the same result, it still hangs whenever you try to build `webpack` and the DashboardPlugin is passed in. Are you using the latest 2.2.1 webpack? @fancyfootowork

**Comment by sethbro** [comment]
@MarkPieszak I'm still on webpack 1.12.0

This is what I wound up with in config:
```
var argv = require('yargs').argv;
var DashboardPlugin = require('webpack-dashboard/plugin');

if (argv.watch) {
    config.plugins.unshift(
        // Display build info using webpack-dashboard (https://github.com/FormidableLabs/webpack-dashboard)
        new DashboardPlugin({ port: 3033 })
    );
}
```

And "start task":
```
"start": "NODE_PATH=$NODE_PATH:./src:./lib NODE_ENV=development webpack-dashboard -p 3033 -c cyan -- webpack --watch",
```

**Comment by tschaub** [comment]
I've proposed a fix for this in #215.  In the meantime, here is a workaround:

```js
// webpack.config.js
const DashboardPlugin = require('webpack-dashboard/plugin');

const devConfig = {
  // rest of config as usual...
  plugins: [
    {
      apply: function() {
        const dashboard = new DashboardPlugin({color: 'cyan'});
        dashboard.apply.apply(dashboard, arguments);
      }
    }
  ]
};
```
